libraries/AP_Arming/AP_Arming.cpp

```diff
@@ -430,15 +435,46 @@ bool AP_Arming::gps_checks(bool report)
             return false;
         }
 
-        // check AHRS and GPS are within 10m of each other
         const Location gps_loc = gps.location();
         Location ahrs_loc;
         if (AP::ahrs().get_position(ahrs_loc)) {
+            // check AHRS and GPS are within 10m of each other horizontally
             const float distance = gps_loc.get_distance(ahrs_loc);
             if (distance > AP_ARMING_AHRS_GPS_ERROR_MAX) {
                 check_failed(ARMING_CHECK_GPS, report, "GPS and AHRS differ by %4.1fm", (double)distance);
                 return false;
             }
+            // check AHRS and GPS are close vertically
+            int32_t gps_alt_cm;
+            if (!gps_loc.get_alt_cm(Location::AltFrame::ABSOLUTE, gps_alt_cm)) {
+                check_failed(ARMING_CHECK_GPS, report, "Failed to retrieve GPS altitude");
+                return false;
+            }
+            int32_t ahrs_alt_cm;
+            if (!ahrs_loc.get_alt_cm(Location::AltFrame::ABSOLUTE, ahrs_alt_cm)) {
+                check_failed(ARMING_CHECK_GPS, report, "Failed to retrieve AHRS altitude");
+                return false;
+            }
+            const uint32_t distance_z_cm = abs(gps_alt_cm - ahrs_alt_cm);
+            if (distance_z_cm > AP_ARMING_AHRS_GPS_ERROR_Z_MAX_CM) { //
+                check_failed(ARMING_CHECK_GPS, report, "GPS and AHRS differ in altitude by %.1fm", distance_z_cm/100.0f);
+                return false;
+            }
+            if ((checks_to_perform & ARMING_CHECK_ALL) || (checks_to_perform & ARMING_CHECK_BARO)) {
+                int32_t gps_alt_rel_cm;
+                if (!gps_loc.get_alt_cm(Location::AltFrame::ABOVE_HOME, gps_alt_rel_cm)) {
+                    check_failed(ARMING_CHECK_GPS, report, "Failed to retrieve GPS home-relative altitude");
+                    return false;
+                }
+#if !APM_BUILD_TYPE(APM_BUILD_APMrover2) // Rover doesn't get altitudes correct
+                // also check that the Baro and GPS align:
+                const float baro_distance_z_cm = fabsf(gps_alt_rel_cm - AP::baro().get_altitude()*100.0f);
+                if (baro_distance_z_cm > AP_ARMING_BARO_GPS_ERROR_Z_MAX_CM) {
+                    check_failed(ARMING_CHECK_GPS, report, "GPS and BARO differ in altitude by %.1fm", baro_distance_z_cm/100.0f);
```

Comment:
The motivating log for this check would have been clipping the status text string to the GCS which would have been a bit confusing as it would have lost the unit for the user.

Reply:
Shortened it.